### ISSUE TITLE
Angular 5: NPM Link - model/index.ts is not part of the compilation. 

### ISSUE THREAD

**Issue by thaoula** [angular/angular-cli#8284]
### Bug Report or Feature Request (mark with an `x`)
```
- [x] bug report -> please search issues before submitting
- [ ] feature request
```

### Versions.
Angular CLI: 1.5.0
Node: 8.7.0
OS: darwin x64
Angular: 5.0.0

### Repro steps.
We have a the following folder structure in a git mono-repo

/api (Express Node App)
/model (NPM package using NPM Link, contains mostly typescript interfaces)
/web (Angular)

When using Angular 4.46 and Angular CLI 1.5.rc2 and we were able to make reference to the interfaces defined in the model folder using npm link and using the following example 
`import { entity } @app/model`

We use ng serve to start the app.

After upgrading Angular to V5 and Angular CLI to 1.5.0 we now get the following error -

### The log given by the failure.
Module build failed: Error: /Users/thaoula/Projects/platform/model/index.ts is not part of the compilation. Please make sure it is in your tsconfig via the 'files' or 'include' property.

The occurs for JIT and AOT builds.

### Desired functionality.
Would like be able to build the app using Angular 5

Thanks,
Tarek

**Comment by rolaveric** [comment]
I can report a similar issue.
Key differences:
1. I'm not using `npm link`, I've just installed another git repo via npm with typescript source (ie. original `.ts` files).
2. The error I'm getting is: `node_modules/other-repo/src/lib/index.ts is not part of the compilation output. Please check the other error messages for details.`

Was previously working fine with Angular 4.4.6 and Angular CLI 1.4.7

I'll try to create a minimal repo to reproduce the issue once I get home.

**Comment by suresh2018** [comment]
yes i m also faced this issue .
 angular 5 rc 9 and clic 1.5.0

**Comment by xiongyayun428** [comment]
me too

**Comment by ChristianSchroedel** [comment]
Same here - without npm link, but with dependencies on typescript libraries in other git repositories

**Comment by diagramatics** [comment]
I've been testing with several version combinations to flush this out and it seems like it's the CLI causing it. I'm running v1.5.0 with Angular on v5.0.0 final and a similar error shows. Rolling the CLI back to rc8 fixes it.

Then I reverted https://github.com/angular/angular-cli/commit/9b43a9c30420a6fae10de32af90d27c1c0c4d87b manually and the error subsides.

**Comment by rolaveric** [comment]
I've found that I only get the problem with `ng serve` - not with `ng build`.
Specifically: `ng build --aot --prod --sourcemaps --stats-json --delete-output-path=false --progress=false`

**Comment by rolaveric** [comment]
I can confirm what @diagramatics said about rolling back to rc8 working, but I didn't get the same joy from reverting 9b43a9c
I also noticed that I get the same error when running `ng test` (ie. `X is not part of the compilation output.`) - even with rc8.

**Comment by filipesilva** [comment]
Hey all, this is a side effect of the stricter tsconfig as described in https://blog.angular.io/version-5-0-0-of-angular-now-available-37e414935ced#148d.

The tsconfig file is what determines what TS files are compiled. The error you are getting means that the files mentioned are NOT being compiled.

Before we used to compile all files, essentially ignoring what tsconfig listed. This was a bug, because if you don't include a file in the compilation then it should not be compiled.

By default `./src/tsconfig.app.json` will only pick up files inside `src/`. If you have a file outside source, it won't be picked up. But you can add more files to the tsconfig via either `files` or `include`.

So for @thaoula's case, you probably want to add this the files in `../models` to your `tsconfig.app.json` and also to your tsconfig.spec.json)

```
{
  "extends": "../tsconfig.json",
  "compilerOptions": {
    // ...
  },
  "include": [
    "../models/**/*.ts"
  ],
  "exclude": [
    // ...
  ],
}
```

@rolaveric's case is slightly different. You have TS files in your `node_modules`. This really goes against how libraries should be packaged: libraries should never ship their source `.ts` files.

The reason for this rule is that the TypeScript version your app isn't necessarily the same as the TS version your library uses. Different versions of TS can produce different output, and give different errors for the same input. They don't even support the same language features. So packaging libraries with TS sources will always break someone's project.

So we don't really support using incorrectly packaged libraries with TS sources. It's something you do at your risk.  

That being said, maybe you can try adding it to the include array. But I can't guarantee that will work in the future because it's still and incorrectly packaged library.

Regarding AOT: as far as I can tell, AOT is still incorrectly compiling TS that is not included in the tsconfig. I'll bring this up with the compiler team.

**Comment by MakesNine** [comment]
We have the same problem described by @thaoula , but for a file that is located inside the normal tree structure created by the angular-cli. The error occurs with angular 5.0.0 final and angular-cli 1.5.  We can "ng serve" the app with angular 5.0.0 and angular-cli 1.4.4.

The file that is not part of the compilation in our project is located at:
src/app/core/stateManagement/Models/genericCourse.ts

We never included or excluded files using tsconfig.json so it doesn't seem that the problem comes from the stricker tsconfig usage describeb above.  To make sure, I created an empty project using angular-cli 1.5.0 and replaced the tsconfig.json and tsconfig.app.json files in our project with the ones created by the new cli and it didn't solve the problem.

Another thing we have in commun with @thaoula is that the file that is not included in the compilation only contains interface and class definitions.  So there might be something with that.

**Comment by filipesilva** [comment]
@MakesNine that sounds like a different issue altogether, and one that we thought was fixed. Could you give me a repro of that please?

**Comment by KevinFernandes** [comment]
We are having a similar issue, as well, after upgrading.  It is complaining about all of our library packages under node_modules with this same exact error (all packages are installed with NPM so no links).  Are we going to have to "include" each package separately in the tsconfig file?  This would be a nightmare if its true, now and for maintenance in the future.  Is there another solution or something else we can do first before using this new version which obviously breaks us completely?

**Comment by MakesNine** [comment]
@filipesilva I have been trying to reproduce the problem in a small project, but I haven't succeded so far.   I will keep trying to isolate the problem to find the source of it.

Here's the exact error I'm getting: 

ERROR in ./src/app/core/StateManagement/models/genericCourse.models.ts
Module build failed: Error: D:\Projects\LtiisWeb\src\app\core\StateManagement\models\genericCourse.models.ts is not part o
f the compilation. Please make sure it is in your tsconfig via the 'files' or 'include' property.
    at AngularCompilerPlugin.getCompiledFile (D:\Projects\LtiisWeb\node_modules\@ngtools\webpack\src\angular_compiler_plug
in.js:624:23)
    at plugin.done.then (D:\Projects\LtiisWeb\node_modules\@ngtools\webpack\src\loader.js:467:39)
 @ ./src/app/+generic-courses/generic-course.component.ts 23:0-84
 @ ./src/app/+generic-courses/generic-courses.module.ts
 @ ./src/$$_lazy_route_resource lazy
 @ ./node_modules/@angular/core/esm5/core.js
 @ ./src/main.ts
 @ multi webpack-dev-server/client?http://0.0.0.0:0 ./src/main.ts

**Comment by filipesilva** [comment]
@MakesNine so if you try having a file with just interfaces it's still fine in a new project? Hm... I've heard of problems with casing. Are your imports perhaps using the wrong casing? Does it still happen if you add a non-interface export to that file? That seems to be a lazy route, perhaps it's related. Can you try making that route not lazy?

@KevinFernandes this definitely shouldn't happen to all library packages, unless somehow you're just using packages that weren't packaged well. Can you give me examples of a couple of these packages? I can have a look.

**Comment by MakesNine** [comment]
@filipesilva yest, I can build a new project and import a file with only interfaces from a lazy loaded module.  I will double check the casing and try the original app with a non lazy route for the module where the error originates.

**Comment by MakesNine** [comment]
@filipesilva I have added an import and exported a function from the file refusing to be part of the compilation and it didn't help.
I have changed all the routes, in app-routing.module.ts,  that have components importing type definitions from that file so that they wouldn't be lazy loaded and the app compiled, but didn't work.  I had forgotten to import the modules, in app.module.ts', that contained the components now used in the non lazy loaded routes.  Once I add the required imports in app.module.ts, I get the same problem that I get with lazy loaded routes.  I have also checked casing and all seems ok on that front.

**Comment by rolaveric** [comment]
Thanks @filipesilva - that explains my issue.
I tried explicitly including my internal libraries using `include`, but that's giving me other grief (picking up on test suites, despite using `exclude`). So I'll do it properly and have them expose compiled code.
I knew it was something I should do, but with them only being used internally by the company, I let it slide.

Thanks again to you and all the contributors for your hard work on Angular CLI - it's an excellent tool and very much appreciated.

**Comment by stuartkuentzel** [comment]
I was having the same issue after updating. Like some other people, the ts files were all part of the standard CLI structure tree. Turned out it was a case sensitivity issue for me. I had accidentally written something like...

`import { Comment } from './Comment'` while in other components `import { Comment } from './comment'`. Running a normal build would throw me a helpful warning before erroring out, but building for prod only showed me the "Please make sure it is in your tsconfig via the 'files' or 'include' property" error. Fixed the casing issue and prod builds for me fine.

**Comment by MakesNine** [comment]
I finally got my app to compile.  It turned out that it was a casing issue indeed.
The error I was getting was: 
ERROR in ./src/app/core/StateManagement/models/genericCourse.models.ts

And the solution is to change to related import to:
 ./src/app/core/**s**tateManagement/models/genericCourse.models.ts

Curiously, I have other files in that folder that were also imported with StateMangement in the path and the compiler didn't complained.

Thansk for your input @filipesilva

**Comment by prasadkadiyala** [comment]
We got the same problem when importing the modules.

**Comment by filipesilva** [comment]
@MakesNine that is weird indeed, and I'm not really sure why that happens... I expected all bad paths to throw an error somehow. Will keep an eye out for more such casing reports.

**Comment by KevinFernandes** [comment]
@filipesilva I see hundreds of errors so its hard to sort through them.  From what I can see at the moment it's all of "my" library packages, packages that we build and publish to a local repository.  We package all of the .ts files as well as a webpacked bundle file.

When we were building with SystemJS it made sense that we had to explicitly specify all of the bundles to include from node_modules, but with the CLI it seemed to just work without any extra specifications.  It made our lives a thousand times easier.  But now it seems what we thought was a feature was actually a bug in CLI and possibly something that we are doing wrong (and have been for more than a year now).

So what exactly is the problem?  Is it that the CLI was compiling the .ts files that we had in our library packages and so did not need the bundles?  If that is the case then how do we now tell the CLI to use the bundle?  Is it something to do differently when packaging the library bundles or is it some angular-cli.json configuration we have to change.

Sorry, up until now we had though we had finally figured out this JavaScript eco system, but now it seems we still have some things to learn.

**Comment by filipesilva** [comment]
@KevinFernandes ok so that it happens on your local libs makes sense considering the other reports in this issue. Yes, before it seems like the CLI was compiling the TS in your local libraries, which was just a ticking time bomb really.

The way to address this in your codebase really depends on how you build your libs... Do you use an existing library to do it? I think a lot of users use https://github.com/dherges/ng-packagr and seem to be happy with it. I haven't myself though.

**Comment by emazv72** [comment]
I can report a similar issue:

I have a standard, cli generated, folder structure ( Windows platform )
I simply share some modules between different applications at the source level.

Under src there are the following folders:

- app
- packages (  simlink )

Under packages:

- combo-components
- ....

Each folder is a module ( proper module.ts )  and has an index.ts  for the exports.

tsconfig.json looks like this:

```
"compilerOptions": {
    "baseUrl": "src",
    "paths": {
      "@combo-components/core": [
        "packages/combo-components/src"
      ],
      "@catol/core": [
        "packages/catol-core/src"
      ]
    }
```

An import statement looks like this:
`import { DataSourceResult, NotificationService } from '@combo-components/core';`

It worked fine till angular 4.x and cli 1.4.x
After upgrading to 5.0 & cli 1.5.0, `ng build`  fails with the following error:

ERROR in ../catol/combo-components/src/index.ts
Module build failed: Error: C:\Users\...\workspace\catol\combo-components\src\index.ts is not part of the compilation. Please make sure it is in your tsconfig via the 'files' or 'include' property.
    at AngularCompilerPlugin.getCompiledFile (C:\Users\...\workspace\playground\node_modules\@ngtools\webpack\src\angular_compiler_plugin.js:624:23)
    at plugin.done.then (C:\Users\...\workspace\playground\node_modules\@ngtools\webpack\src\loader.js:467:39)
 @ ./src/app/app.module.ts 13:0-63
 @ ./src/main.ts
 @ multi ./src/main.ts

Casing checked, also added the include property suggested by @filipesilva 

```
{
  "extends": "../tsconfig.json",
  "compilerOptions": {
    // ...
  },
  "include": [
    "./packages/**/*.ts"
  ],
  "exclude": [
    // ...
  ],
}
```

The error changes slightly:

ERROR in ./src/main.ts
Module build failed: Error: C:\Users\...\workspace\playground\src\main.ts is not part of the compilation output. Please check the other error messages for details.
    at AngularCompilerPlugin.getCompiledFile (C:\Users\...\workspace\playground\node_modules\@ngtools\webpack\src\angular_compiler_plugin.js:629:23)
    at plugin.done.then (C:\Users\...\workspace\playground\node_modules\@ngtools\webpack\src\loader.js:467:39)
    at process._tickCallback (internal/process/next_tick.js:109:7)
 @ multi ./src/main.ts
ERROR in ./src/polyfills.ts
...

Adding main&polyfill to the include path doesn't help either.

**Comment by zhekaus** [comment]
The same issue with fresh-created app using `ng new`
Angular CLI: 1.5.0
Node: 6.10.3
OS: win32 x64
Angular: 5.0.0

**Comment by scottieslg** [comment]
I was having this issue and it was a case issue.

I was trying to do this:

`import { IJobModel, JobModel } from "./**_jobModels_**/job.model";`

which should be 

`import { IJobModel, JobModel } from "./**_JobModels_**/job.model";`

**Comment by sandangel** [comment]
have you tried with --preserve-symlinks flag? https://github.com/angular/angular/issues/20091

**Comment by emazv72** [comment]
@sandangel Thank you that worked for me, no need to add the 'include' property in tsconfig.json

After upgrade, just run

`ng serve --preserve-symlinks`

if you have symlinks in the source tree.

@filipesilva It seems that the default behavior has been reverted since cli 1.4. It would be nice to have a per environment option in tsconfig.app.json/.angular-cli.json like this:

```
{
  "preserveSymlinks": {
    "dev": true,
    "prod": false
  }
}
```

**Comment by HashSix** [comment]
I'm using a 3rd party library from Google called api-ai-javascript which is installed in the node_modules folder, using `ng serve --preserve-symlinks` does not work and neither does updating my `angular-cli.json` file, I have to explicitly  update my tsconfig file with `include` in order to overcome the following error ```Module build failed: Error: C:\Users\Alex.Susini\Documents\Projects\ManagerV2Desktop\node_modules\api-ai-javascript\index.ts is not part of the compilation output. Please check the other error messages for detai
ls.
    at AngularCompilerPlugin.getCompiledFile (C:\Users\Alex.Susini\Documents\Projects\ManagerV2Desktop\node_modules\@ngtools\webpack\src\angular_compiler_plugin.js:629:23)
    at plugin.done.then (C:\Users\Development\Documents\Projects\ManagerV2Desktop\node_modules\@ngtools\webpack\src\loader.js:467:39)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
 @ ./src/app/services/bot/bot.service.ts 12:0-48
 @ ./src/app/components/common/app-common.module.ts
 @ ./src/app/app.module.ts
 @ ./src/main.ts
 @ multi ./src/main.ts```

Only after updating to Angular 5.0.0 and Angular-cli 1.5.0 did the issue start occurring.

**Comment by SamD** [comment]
I'm also seeing the issue after upgrading to Angular 5 and cli 1.5.0 with a third-party logger module that had no issues previously.

Tried suggested methods posted here still the same error.

`ERROR in ./node_modules/angular2-logger/app/core/level.ts
Module build failed: Error: node_modules/angular2-logger/app/core/level.ts is not part of the compilation. Please make sure it is in your tsconfig via the 'files' or 'include' property.
`

**Comment by firejar710** [comment]
I am have the same error, although mine came about after I imported the 'Input' class from '@angular/core'.

`import {Component, OnInit, ViewEncapsulation, Input} from '@angular/core';`

After I used the `@Input()` annotation, I had this problem. removing it helped. I wonder if it's a casing issue like some of the previous posts, although I didn't change where the import was pointing. What could it be?

**Comment by michaelwaihenya** [comment]
@scottieslg pointed me in the right direction. (Typo in path casing). Additionally I had several backticks (`) in my compoments that were compiling before.  single quotes (') resolved the issue.

**Comment by little-boy** [comment]
I encountered the problem with a project having a Symfony like architecture.

Meaning there is : 
./tsconfig.json
./src/tsconfig.app.json
./src/SomeFolder/app/*.ts...

There are dependencies between each bundles.

One of the dependencies was thrown in "...is not part of the compilation. Please make sure it is in your tsconfig via the 'files' or 'include' property"

I solved the problem by including every ts files as suggested by @filipesilva **but in a different order**.

=> Bundle B needed by Bundle A & C
so tsconfig.app.json like : 
`{
    "include": [
        "pathToBundleB/**/*.ts",
        "pathToBundleA/**/*.ts",
        "pathToBundleC/**/*.ts",
    ]
}`
instead of : 
`{
    "include": [
        "pathToBundleA/**/*.ts",
        "pathToBundleB/**/*.ts",
        "pathToBundleC/**/*.ts",
    ]
}
`

That's kinda weird. Looks like @ngtools/webpack/src/angular_compiler_plugin.js:624:23), the function this._compilerHost.fileExists is looking for bundles loaded before, and not after..

**Comment by SaltyDH** [comment]
I'm getting this error for files that are referenced in our app.module.ts. Its almost like this file is treated differently in some way because its the module referenced in the AngularCompilerPlugin options??? I've tried manually adding the module to the tsconfig files section but that hasn't resolved it.

**Comment by sbkpilot1** [comment]
We are having the same issue.  We have a private repo where we have published a library that we use.  We have this lib as a dependency in our client, when compiling in Angular 5 we get the following error but ONLY in JIT mode, if we use a ng serve --aot or ng build --aot or --prod it works fine.  Not sure what is the issue here.  

We tried using includes in tsconfig, while that solves this issue it generates another different error that my json module declarations are not found which is specified in typings.d.ts (with a declare module "*.json").

```
ERROR in ./node_modules/@myscope/services-lib/index.ts
**Module build failed: Error: [path]\node_modules\@myscope\services-lib\index.ts is not part of the compilation output. Please check the other error messages for details.**
    at AngularCompilerPlugin.getCompiledFile ([path]\node_modules\@ngtools\webpack\src\angular_compiler_plugin.js:629:23)
```

**Comment by rolaveric** [comment]
I've managed to resolve the issue I was having by, as per @filipesilva's advice, compiling the TS code from my library and configuring it so my downstream projects could consume it.
I did have a lot of trouble with this though, so I want to share some details in the hope they'll help others with the same frustrations.

Most of the documentation and discussion available frames these repos we're installing via npm as "libraries" in the open source sense of the term: The code is on github, the artifacts are published to npm, and they must be built for general consumption (ie. accounting for different projects with different ES targets).
My situation (which is not unique, from the sound of things) is different in that my "library" is for internal company use only.
To help separate this use case from the "open source library" scenario, I'm going to call what I've got a "shared codebase".

Your starting point should be @filipesilva's [`angular-quickstart-lib` repo](https://github.com/filipesilva/angular-quickstart-lib).
If, like me, you've got a "shared codebase" then you don't need everything from here. The key things you need are:
1. To inline resources before you compile them (ie. `templateUrl` becomes `template)
2. To compile the TS code using `ngc`, with appropriate `tsconfig.json` settings
3. To add `module` and `typings` properties to your `package.json`, pointing at the `index.js` and `index.d.ts` of your compiled code.

What you don't need is: Bundling (Just use the raw `*.js` files), and multiple compile targets (use ES5 or ES2015.
So you can strip down a lot of the `build.js` file (removing the need for a fair few extra npm dependencies) and ignore references to `flatModule` in the `tsconfig.lib.json`.

I hope that helps.

**Comment by jjoao07** [comment]
This really stressful!

Myself  and my team, we have put all energy believing and trusting Angular, 
that's this time things will be different not like from angularJs to angular 2.

And we could jump a long time from angularJs to ReactJs or Aurelia or maybe VueJs,
but the love and trust we had it keeps as loving  Angular and believing that we won't have same stress like we had from angularJs to angular 2.

Now we already with big projects out there, using third-party open source libraries, which till angular 4.3, everything was fine.

But now in angular 5 with the CLI 1.5, we were expecting the upgrade to be smooth and not having this kind of issue.
So now we're stuck and hoping that the creators of the libraries they update the package.

And the questions come again, should we keep believing that angular will have the success like angularJs had on the past, or now is time to move on trying something different because this is a nightmare, and each version got different stories, and we must signup new courses to learn over.

I'm sorry for my comments, but this how it feels like.

![image](https://user-images.githubusercontent.com/19816874/32688805-78876cbc-c6e0-11e7-961c-b8e59954ba07.png)

**Comment by deebloo** [comment]
@jjoao07 the issue isn't an angular issue it is a typescript issue. To solve it for linked libraries that have typescript shipped in them you can add that to your tsconfig "includes" property. So if you have a folder in your node modules called "my-lib" you can add. The typescript compiler by default does not see the typescript code in node_modules by default (and it shouldn't). I hope this helps. 

Also sorry if I am repeating what someone else already said.

```JSON
{
    includes: ['node_modules/my-lib']
}
```

**Comment by sbkpilot1** [comment]
@deebloo if the Typescript compiler shouldn't be looking at ts code in the node modules then perhaps adding an includes is not the best solution for this?  Should we be packaging our libraries differently?

**Comment by deebloo** [comment]
@sbkpilot1 overall yes. if you aren publishing a module you should only be distributing javascript and  the generated type definitions. The solution I mentioned is only a workaround

**Comment by akkumar** [comment]
add the following to **.npmignore** (in addition to  'includ'ing in tsconfig.json, of course)
```text
# Ignored files
*.ts
!*.d.ts
```

**Comment by malaba03** [comment]
hello guys,
I had the same issue in my project (Angular 5.0.0).
I solved by deleting the file which causes the problem.
After i created it again and pasting the same code step by step in order to make sure what part of the code cause the problem.
Then everything work.
Hope my solution will help you.

@malaba

**Comment by JulienBourgain** [comment]
@filipesilva I understand your arguments, but why release a breaking change in a minor version ?!

It's better to add a warning in this version and correct this behaviour in 2.0.0

**Comment by criollapp** [comment]
If you are working with multiple libraries with npm link under a scope and you have something like the following in your tsconfig:

`"paths": {
      "@my-libs/*": ["../node_modules/@my-libs/*"]
    }`

The error will be present. Remove it

**Comment by Jesu26** [comment]
Hi ,
I am also facing the same problem.

while compiling the solution via ng serve -o after angular 5 update this issue occurs.

i have used npm link to link all my modules.

ERROR in ./platform/userModule/app/app.module.ts
Module build failed: Error: E:\temp\trunk\platform\userModule\app\app.module.ts is not part of the compilation. Please make sure it is in your tsconfig via the 'files' or 'include' property.
at AngularCompilerPlugin.getCompiledFile (E:\temp\trunk\node_modules@ngtools\webpack\src\angular_compiler_plugin.js:624:23)
at plugin.done.then (E:\temp\trunk\node_modules@ngtools\webpack\src\loader.js:467:39)
at 
@ ./src/$$_lazy_route_resource lazy
@ ./node_modules/@angular/core/esm5/core.js
@ ./src/main.ts
@ multi ./src/main.ts

**Comment by usernameisMan** [comment]
i also met the  question , 
`
ERROR in ./src/app/hero.ts
Module build failed: Error: F:\lenleeWorkProject\angular\angular-tour-of-heroes\src\app\hero.ts is not part o
f the compilation. Please make sure it is in your tsconfig via the 'files' or 'include' property.
    at AngularCompilerPlugin.getCompiledFile (F:\lenleeWorkProject\angular\angular-tour-of-heroes\node_module
s\@ngtools\webpack\src\angular_compiler_plugin.js:624:23)
    at plugin.done.then (F:\lenleeWorkProject\angular\angular-tour-of-heroes\node_modules\@ngtools\webpack\sr
c\loader.js:467:39)
    at <anonymous>
 @ ./src/app/hero-detail/hero-detail.component.ts 11:0-31
 @ ./src/app/app.module.ts
 @ ./src/main.ts
 @ multi webpack-dev-server/client?http://0.0.0.0:0 ./src/main.ts
`
============================
i am check all the  hero.ts   reference， The discovery problem appears on the file path     
is  Uppercase the Hero Instead of hero 。

so why path error not prompting？ 

Instead, it returns to normal after manually triggering the save
============================
Angular CLI: 1.5.0
Node: 8.4.0
OS: win32 x64
Angular: 5.0.1
... animations, common, compiler, compiler-cli, core, forms
... http, language-service, platform-browser
... platform-browser-dynamic, router

@angular/cli: 1.5.0
@angular-devkit/build-optimizer: 0.0.33
@angular-devkit/core: 0.0.20
@angular-devkit/schematics: 0.0.35
@ngtools/json-schema: 1.1.0
@ngtools/webpack: 1.8.0
@schematics/angular: 0.1.3
typescript: 2.4.2
webpack: 3.8.1

**Comment by bzah** [comment]
Hello, I came across the same issue using angular 5.0.1 and CLI 1.5 :

> Module build failed: Error: xxx\ng4-blog-frontend\src\app\models\post.ts is not part of the compilation. Please make sure it is in your tsconfig via the
'files' or 'include' property.

At first I had a case issue that I solved and it worked fine for a while but the error came back somehow later.
I tried what @sandangel suggested by using the flag `--preserve-symlinks` but it did not worked.

Finally, I found another issue (#20091) where @stephanecot suggested the use of `--aot` and it works fine, although I have no idea why...
As the flag suggest, it use the[ Ahead of time compilation](https://angular.io/guide/aot-compiler) which can be unwanted. 

I hope it will help someone.

**Comment by khalid-halo** [comment]
Tried workarounds mentioned in this thread, dosen't seem to help. Still getting the same error.

`ERROR in ./node_modules/angular2-isotope/index.ts`

How can I compile the library locally to continue development for the time being?

**Comment by misaizdaleka** [comment]
What I did with my library is I moved it to the src folder and included it from there. Not a clean solution, but it'll work until the Angular guys figure this one out.

**Comment by nueko** [comment]
FOR ME,
The issue comes up when the project stands on a junction directory or symlink.

**Comment by rychkog** [comment]
I have same symptoms like @nueko has. When one of my dependencies is linked via symlink and I invoke `ng build --prod --build-optimizer` - then it fails. 
`--preserve-symlinks` helps to fix this particular case:
`ng build --prod --build-optimizer --preserve-symlinks`

**Comment by milosmijakovac** [comment]
This is my problem:
Angular-cli :  1.5.4
npm  5.5
node 8.9

````ERROR in ./src/app/hero.ts
Module build failed: Error: C:\...\src\app\hero.ts is not part of the compilation. Please make sure it is in your tsconfig via the 'files' or 'include' property.
    at AngularCompilerPlugin.getCompiledFile (C:\...\node_modules\@ngtools\webpack\src\angular_compiler_plugin.js:620:23)
    at plugin.done.then (C:\...\node_modules\@ngtools\webpack\src\loader.js:467:39)
    at <anonymous>
 @ ./src/app/hero-details/hero-details.component.ts 11:0-31
 @ ./src/app/app.module.ts
 @ ./src/main.ts
 @ multi webpack-dev-server/client?http://0.0.0.0:0 ./src/main.ts````

**Comment by HollisTech** [comment]
adding: to the defaults section of  .angual-cli.json 
,
    "build": {
      "preserveSymlinks": true
    }

corrects this malfunction. 100% reproducible with generated app.

**Comment by nueko** [comment]
@HollisTech, or just 
```bash
ng serve --preserve-symlinks
```

**Comment by milosmijakovac** [comment]
@HollisTech
 it shows same warning! : (

**Comment by angryanatole** [comment]
I am having the same problem.

`ng serve --preserve-symlinks` didn't help (I didn't have any typings), but `ng serve --aot --preserve-symlinks` did actually help.

Angular CLI 1.5.5

**Comment by 22ryann** [comment]
Yep, same issue here with upgrading to 5.1, never had this issue before now.

My include ended up looking like this to bring everything in I needed for compilation.

`  
"include": [
    "**.ts",
    "app/**.ts",
    "app/**/**.ts",
    "../node_modules/angular2-semantic-ui/components/**/**.ts"
  ]

`

**Comment by mattytommo** [comment]
We're using the "shared codebase" approach @rolaveric is using and I managed to get it to compile and run fine by adding the following include to the tsconfig.json:

    "include": [
        "src/**/*",
        "node_modules/{sharedcodebasename}/**/*"
    ]

That all works great, however if I try and run `ng test` I end up with the same build errors that I got before adding our shared codebase to the includes (Module build failed: "{sharedcodebasename}/file.ts" is not part of the compilation output.

Now @filipesilva I appreciate you saying this isn't the "proper way to do an NPM package", but for my case (and what looks like a few others in this thread), we have internal company "packages" that we don't want pushing to NPM (we push them to our private Nexus repo) and publishing typescript files as a package is exactly what we need. Is there any way that this can be resolved? I understand the rationale behind the change in compilation behaviour, but it seems to be causing many unintended headaches.

**Comment by rkrisztian** [comment]
@mattytommo, instead of editing `tsconfig.json` like that, you can add the include(s) into `tsconfig.app.json` + `tsconfig.spec.json` (prefixed with `../` of course).

**Comment by mattytommo** [comment]
@rkrisztian I must have tried it in those files 100 times, but I don't think I actually prefixed with ../ because adding that  to the spec.json file works!!

So, essentially, as long as you put the include within the spec.json (for the tests) and the tsconfig.json (or .app.json), the build and tests work. Thanks again rkrisztian.

**Comment by estebanluengo** [comment]
I don't think is a problem of angular-cli. I compile my projects directly with webpack and my own config files and I have the same problem if a use AOT option with Angular 5.x.x but it's perfect in Angular 4.x.x. But if I don't use AOT option the compilation is fine even in Angular 5.x.x. Maybe the problem is in the Angular Ahead-of-Time Webpack Plugin: @ngtools\webpack 

ERROR in ./src/app/common/data/reserva.ts
Module build failed: Error: G:\Proyectos\BITCITA\bitcita-front\src\main\frontend\src\app\common\data\reserva.ts is not part of the compilation. Please make sure it is in your tsconfig via the 'files' or 'include' property.
    at AngularCompilerPlugin.getCompiledFile (G:\Proyectos\BITCITA\bitcita-front\src\main\frontend\node_modules\@ngtools\webpack\src\angular_compiler_plugin.js:643:23)
    at plugin.done.then (G:\Proyectos\BITCITA\bitcita-front\src\main\frontend\node_modules\@ngtools\webpack\src\loader.js:467:39)
 @ ./src/app/recursos/component/vacaciones-list.component.ts 21:16-52
 @ ./src/app/recursos/component/recurso-form.component.ngfactory.js
 @ ./src/app/campanyas/campanya.module.ngfactory.js
 @ ./$$_lazy_route_resource async
 @ ./~/@angular/core/esm5/core.js
 @ ./src/vendor-aot.ts

**Comment by rgantla** [comment]
I have faced the same issue and turns out, in my case it was a Casing Issue.

I had an import like so: **import { Employee } from '../../../models/employeeType.model';** in more than one component.

One of the components that had this import had a different casing in the file path like **import { Employee } from '../../../models/EmployeeType.model'**; 

Changing the file path for the import to be consistent across the application fixed the issue for me.

**Comment by milosmijakovac** [comment]
@rgantla  wooww..I wondered how long it would take (centuries) to figure out where I was wrong .
It was exactly Casing issue!
Very thanks!

**Comment by estebanluengo** [comment]
@rgantla OMG!!. That was exactly the problem. It was a Casing Issue. Thanks!!

**Comment by sbley** [comment]
In my case it was not exactly a casing issue but also an import that had to be fixed. 

Using [ngx-clipboard](https://www.npmjs.com/package/ngx-clipboard) I had to change the import from 
`import { ClipboardModule } from 'ngx-clipboard/index';`
 to 
`import { ClipboardModule } from 'ngx-clipboard/dist';`. 

Hope that helps someone in the same situation.

**Comment by zhekaus** [comment]
I have exactly the same problem if I run `ng serve` from symbolically linked folder (`~/src/...` in my case).
However, if I run `ng serve` from the real folder (`/c/src/...` in my case), everything is fine.

**Comment by bnohad** [comment]
This issue caused most of the teams in my Enterprise company to drop all their Angular X projects & switch to React based application (because we have a huge shared code-base & there is no easy way to export Angular projects as NPM libraries).

I actually liked Angular that's too bad :\ hope this will be solved soon enough before the point of no return.

**Comment by ranbuch** [comment]
I had the same problem in one of my ```npm modules```.
I've solved it by adding a declaration files ```*.d.ts```.
I think you should open an issue to the ```npm module``` that your using.

**Comment by txinfo** [comment]
I have a library project that I (up until the switch to Angular 5) was linking into my consumer project via 'npm link' on the library's src directory. With Angular 4, running 'ng serve' and altering a source file of the library would trigger a recompile against the new library code & refresh the page in browser which was very convenient.

As per this ticket and others, this functionality no longer seems to work in Angular 5 due to stricter TypeScript settings.

I have tried two things to get close to the previous behaviour and neither are proving ideal:
 
1. 

npm link to dist instead and set up a watch to re-run ngc on the library as source files change. This rebuilds the library and triggers ng serve recompilation - however ng serve most of the time fails to pick up the new library at the right time (e.g. after the dist files have been deleted but before library is fully built) and sometimes even when successful does not seen to pick up changes.

2.

include the ts files for the library explicitly within the tsconfig file and run ng serve with --preserve-symlinks. This will get the code to compile successfully, however changes to the code outside of the app's root directory does not trigger the watch.

Is there a way to revert to the original behaviour or fix method 2 above to also watch and trigger an recompile from angular-cli when external ts files are changed?

**Comment by sternetj** [comment]
@SaltyDH I had a similar issue but found that it was a result of the version of TypeScript that I had installed.  I had `^2.5.1` in my package.json and thought I was using 2.5.3. Turns out I had 2.5.1 installed and it was causing the issue.

1. I changed my dependency to `2.5.x`, the same as the Angular repo
2. Did an `npm install`
3. Verified with `npm list typescript` that I had 2.5.3 installed
4. Ran `ng serve` and my app built without issue

**Comment by tobat50** [comment]
@rkrisztian I had the issue just with `ng test`. 
Add `../` in the include paths in the `tsconfig.spec.json` resolve my problem. 
It seems the inheritance of the `tsconfig.json` doesn't work for the paths, because the base directory isn't the same.

**Comment by BBaysinger** [comment]
I have an issue with the workaround. My SO question: https://stackoverflow.com/questions/48449948/how-to-import-repo-with-uncompiled-typescript-in-angular-5-x

**Comment by AndreiLucaci** [comment]
Also, if working with ASP.Core 2.0 template, the `src` folder is renamed to `ClientApp` so in that case you should add it to your `tsconfig.json` file:
`
{
  "compilerOptions": {
    // stuff
  },
  "include": [
    "./ClientApp/**/*.ts"
  ]
}
`

**Comment by b-johnse** [comment]
I have a large enterprise app with the "code sharing" setup mentioned above. Essentially there are lazy routes defined that load feature modules, these modules are shells that in turn import from node_modules where the actual feature modules (in module.ts files via index.ts barrels) are pulled in from our internal npme/nexus. Worked like a champ with Ng4 and enabled our distributed development across multiple teams/repos. 

With Ng5, after updating the tsconfig files, it will work in JiT (ng serve and ng build).

AoT will build, but when the app is started a "Uncaught TypeError: ctor is not a constructor" is thrown immediately .

**Comment by stofte** [comment]
Seems silly that we have to include files that are normally part of the build. Currently my whatever is complaining that " \src\main.ts is missing from the TypeScript compilation." but that's the very very first file angular looks at, so that seems strange.

Oh well.

**Comment by plamf** [comment]
@filipesilva 

> You have TS files in your node_modules. This really goes against how libraries should be packaged: libraries should never ship their source .ts files.

What exactly do you mean? We use ng-packagr, which ships *.d.ts files and we thought that is the right way to deliver a package.

**Comment by benelliott** [comment]
@plamf Shipping *.d.ts is fine, it is shipping the original source *.ts files that is not.

**Comment by BBaysinger** [comment]
Well we’re not ‘shipping’ anything. This is an internal library for which we have control over which compliler we use. So packaging for distribution is a unnecessary step. We did figure out how to import it with straight .ts files, as the SO question was answered satisfactorily.

**Comment by sliu4015** [comment]
I came with Angular less than 1 day. I get this error from running cli command ' ng generate service message --module=app'. I removed those files  generated by cli and recovered 'app.module.ts' into it previous look. Tried cli command ' ng generate service message' without flag '--module=app'. Manually modified   'app.module.ts' with "import {**} from '**'" and 'providers:[**]'. The error is gone somehow.  Confused.

**Comment by ORESoftware** [comment]
I have this issue....this issue sucks :( 

this is how I solved it for local development:

https://stackoverflow.com/questions/48797135/missing-ts-files-due-to-npm-link/48798373#48798373

**Comment by kaushikdighe** [comment]
@filipesilva. I am facing the same issue Just to summarize, we have a large project. So 1 project is used as library and sits in the node_modules folder. When on Angular 4 it works fine but when upgraded it is throwing me this error: **Module build failed: XXX file is missing in Typescript compilation. Please make sure it is in your tsconfig via the 'files' or 'include' property.** I have read this whole ticket and specially implemented what you suggested of manually adding it in tsconfig.json. But when I do that it prompts me a different error: **Unknown compiler option 'include' and 'exclude'**. It would be great help if you could get me out of this issue. Thank you in advance. Just to reiterate I have read through the whole blog and tried different solutions it does not seem to work. Again Thank you in advance.

**Comment by kaushikdighe** [comment]
@andi-wr Thank you for the solution but it still does not work. It gives me whole lot of errors. Saying **cannot find name 'describe'; cannot find name 'beforeEach' etc in all the files**. This is a huge list of errors.

**Comment by princeprathap** [comment]
@filipesilva 
I am getting the below error and the third party literary do not have .ts file. can you help

ERROR in ./node_modules/agm-direction/src/agm-direction.module.ts
Module build failed: Error: D:\TandT\TandTSrc\node_modules\agm-direction\src\agm-direction.module.ts is missing from the TypeScript compilation. Please make sure it is in your tsconfig via the 'files' or 'include' property.
The missing file seems to be part of a third party library. TS files in published libraries are often a sign of a badly packaged library. Please open an issue in the library repository to alert its author and ask them to package the library using the Angular Package Format (https://goo.gl/jB3GVv).
    at AngularCompilerPlugin.getCompiledFile (D:\TandT\TandTSrc\node_modules\@ngtools\webpack\src\angular_compiler_plugin.js:656:23)
    at plugin.done.then (D:\TandT\TandTSrc\node_modules\@ngtools\webpack\src\loader.js:467:39)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:169:7)
 @ ./node_modules/agm-direction/index.js 6:9-46
 @ ./src/app/app.module.ts
 @ ./src/main.ts
 @ multi ./src/main.ts

**Comment by tapaz1** [comment]
I don't know if people are still having this problem - here's how I fixed the issue.

I have 2 tsconfig.json files, one at the root of the app, and the other inside the src folder (one level down from the root) named tsconfig.app.json that extends the main tsconfig.json. I explicitly added the package that I needed that wasn't being transpiled by Typescript in the "include" array, and like a lot of people here I was getting the *.spec.ts files included despite having them in "exclude" option, so I removed it from the tsconfig.json. The fix was adding the "exclude" option to the second (extended) tsconfig.app.json file. Doing that fixes the Typescript errors and lets me run ng serve without issue. 

**Here is tsconfig.json:**

{
  "compileOnSave": false,
  "compilerOptions": {
    "baseUrl": "./src",
    "outDir": "./dist",
    "sourceMap": true,
    "declaration": false,
    "moduleResolution": "node",
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "target": "es5",
    "typeRoots": [
      "node_modules/@types"
    ],
    "lib": [
      "es2017",
      "dom"
    ],
  },
  **"include": [
    "./src",
    "node_modules/your_package"
  ]**
}

**And tsconfig.app.json:**

{
  "extends": "../tsconfig.json",
  "compilerOptions": {
    "outDir": "../dist",
    "baseUrl": "./",
    "module": "es2015",
    "types": []
  },
  "**exclude": [
    "test.ts",
    "**/*.spec.ts",
    "../node_modules/your_package/**/*.spec.ts"
  ]**
}

Based on the comments from @filipesilva this may still be a bug but it worked for me. Hopefully this helps anyone facing the same issue.

**Comment by fengerzh** [comment]
@tapaz1 By following your instruction, I got it work. I added a section of `include` in `tsconfig.json` to include the failed module as below:

`"include": ["./src", "node_modules/@jaspero/ng2-select"]`

then build success. Thank you!

**Comment by criollapp** [comment]
The error also happens when you have the url of a main file wrong in the package.json in your library

**Comment by jdthorpe** [comment]
> we used to compile all files, essentially ignoring what tsconfig listed. This was a bug

IMHO, the new behavior of failing when the compiler wasn't asked to compile a file it wasn't supposed to compile in the first place is also a bug.

> we don't really support using incorrectly packaged libraries

NPM link is supported by, well, NPM, so it seems odd that angular doesn't.  In particular, I've got this in in my `.npmignore`: 

    *.ts

So that TypeScript files are not included when I `npm publish`.  Perhaps angular could also respect .npmignore files.  This would resolve the problems that devs who co-develop an angular site and a related npm package have with this new behavior.

**Comment by mpilarcastillejodev** [comment]
add in tsconfig.app   
 "baseUrl": "" 
 for  
 "baseUrl": ""
